### PR TITLE
overhaul generation framework with TemplateSet

### PR DESCRIPTION
**Pull request by jaypipes**
Adds a new `pkg/generate/templateset` containing a `TemplateSet` struct
that makes our template generation extensible. No more hard-coded
template paths or unused templates for Crossplane. Instead, with this
patch, we now have a `pkg/generate/ack` package that contains the
instructions for generating ACK components and a
`pkg/generate/crossplane` that contains instructions for generating
Crossplane components.

Issue #538

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cmd/ack-generate/command/apis.go [hunk 1]
```diff
@@ -84,138 +83,36 @@ func generateAPIs(cmd *cobra.Command, args []string) error {
...
-	if err = writeTypesGo(g, typeDefs); err != nil {
-		return err
-	}
-
-	for _, crd := range crds {
-		if err = writeCRDGo(g, crd); err != nil {
+	apisVersionPath = filepath.Join(optAPIsOutputPath, svcAlias, "apis", optGenVersion)
+	for path, contents := range ts.Executed() {
+		if optDryRun {
+			fmt.Printf("============================= %s ======================================", path)
```
Comment (A-Hilaly) [inline cmd/ack-generate/command/apis.go:104]: nit: all the --dry-run like commands needs a new ligne after printing headers. Thinking something like : `fmt.Printf("============================= %s ======================================\n", path)`

**Review by A-Hilaly — APPROVED** [review]
/lgtm 👍 I left two nit comments :)
I tested this patch locally and everything is working perfectly.

File: cmd/ack-generate/command/apis.go (hunk 1 above)
Comment (jaypipes) [reply cmd/ack-generate/command/apis.go:104]: Nice catch. Fixed.

File: cmd/ack-generate/command/crossplane.go [hunk 2]
```diff
@@ -63,11 +67,40 @@ func generateCrossplane(_ *cobra.Command, args []string) error {
 			return fmt.Errorf("cannot get the API model for service %s", svcAlias)
 		}
 	}
-	var opts []crossplane.GenerationOption
 	cfgPath := filepath.Join(providerDir, "apis", svcAlias, optGenVersion, "generator-config.yaml")
-	if _, err := os.Stat(cfgPath); !os.IsNotExist(err) {
-		opts = append(opts, crossplane.WithGeneratorConfigFilePath(cfgPath))
+	g, err := generate.New(
+		sdkAPI, optGenVersion, cfgPath, cpgenerate.DefaultConfig,
+	)
```
[the remark this answers is not in the document]
Comment (jaypipes) [reply cmd/ack-generate/command/crossplane.go:73]: Fixed. Nice eyeballs. :)

**Review by A-Hilaly — APPROVED** [review]
Neat! let's ship it :)

**Comment by muvaf** [comment]
I just had the time to review the PR and it seems like a big overhaul 🙂 

I see that Crossplane pipeline parts are now more tightly integrated and a few small things got broken, like file paths and names. But overall, it feels good to eliminate those template files. Thanks @jaypipes 

I'll open a PR fixing those path strings, removing a few more unnecessary `.tpl` files from Crossplane alongside some small changes in `pkg/generate/crossplane`. I might be able to squeeze the `get_attributes` templates for Crossplane, too, in that PR.

**Pull request merged**